test/csuite/timestamp_abort/main.c

```diff
@@ -369,6 +383,7 @@ thread_run(void *arg)
         if (use_ts) {
             testutil_check(pthread_rwlock_rdlock(&ts_lock));
             active_ts = __wt_atomic_addv64(&global_ts, 2);
+            WT_PUBLISH(ts_data[td->info], active_ts);
```

Comment:
I still don't understand why we need the barriers.

Reply:
i'm not sure if "undefined behaviour" is the right term here, but leaving these barriers out caused some really spooky behaviour when reading these back and calculating the minimum - i could reliably see it either skipping entries or claiming to have written an entry back when it didn't. the latter case isn't a problem, but the former is.

it also seemed more robust to use these than rely on the "missing a read or write causes a lower timestamp to be seen" behaviour.

Reply:
It's still hard for me to believe the compiler is doing some weird thing in this case.I'm not the expert in this but may be worth asking @keithbostic to have a look.

Reply:
Sure. I'll spend some time confirming that the read/write barriers are helpful, I agree we shouldn't need them and it's possible I just got lucky at the right times when I was testing after adding barriers.

Reply:
@korteland, what is the write-barrier that `WT_PUBLISH` inserts before the assignment supposed to do? That is, if there's a write barrier, then there's reasonably some write the needs to be scheduled in advance of the assignment. What is that write?

Reply:
I've managed to confuse myself here, apologies for the wild goose chase :-(

I wanted a general-purpose memory barrier and wasn't sure if the read+write combo in `hardware.h` was enough, so I went reading and found https://www.kernel.org/doc/Documentation/memory-barriers.txt. It says "General memory barriers imply both read and write memory barriers, and so can substitute for either" which I misunderstood. For whatever reason I ended up thinking a read+write combo could substitute for a full barrier, which is incorrect.

However we're back to square one in a sense, since it was intermittently failing without the barriers but I'm yet to see it fail with them - perhaps I was unlucky when testing it out, but I could still use a hand here. Again, sorry for the misunderstanding.

Reply:
I have seen a similar failure in `format`, and it's been on my list to track down for awhile. Maybe it's time to take a run at it -- maybe there's something I don't understand as well as I think I do. :) 

My suggestion would be to write down exactly what you think the relationship is between the data elements and the constraints around them, then add an assert so that when the error hits you get a core dump. Then, I'd dig through the list of running transactions and see if there's something unexpected going on.

Reply:
@korteland, I just pushed similar work for the `format-mirror` branch (#7662) to remove the code that queried that all_durable timestamp when updating the oldest or stable timestamps. The change is in 4f9b658, if you're interested.

Specifically, I keep track of the last-used commit/prepare timestamp used by each worker thread, and have a function which returns the maximum "not in use" timestamp by walking the array of worker threads, and returning the minimum last-used commit timestamp minus 1. At the same time, I ensure transactions never set a read timestamp, and the application never sets the oldest or stable timestamps, from inside the range of timestamps still in use. That allowed me to get rid of making queries of the `all_durable` timestamp, and it didn't require any barriers (in fact, it got rid of some locking around transaction commit and prepare.

I think this makes sense from an application perspective: admittedly, read timestamps might reasonably be allocated from inside of the in-use timestamp range, but I don't think I'm hurting testing by not doing that, and I think it's pretty unlikely oldest or stable would be set from inside the active timestamp range.

cc: @quchenhao, @kommiharibabu, @agorrod, @sauclovian-wt

Reply:
Thanks for the explanation @keithbostic, I went over that commit and it looks very similar to the `timestamp_abort` changes here. The main difference is that `timestamp_abort` doesn't use a read timestamp and that the incrementing the global time stamp is done on the "consumer" side rather than the "producer" side.

I spent a lot of yesterday and today hammering on it without barriers (and a couple of extra asserts, as you mentioned) and I didn't see another failure, so I'm now thinking I was mistaken with the failures I saw. (Perhaps it was because I was halfway through fixing this branch, although the diff looks fine).

Either way, I think it's either a false alarm or rare enough to not break everything, so if you're OK with the branch now that I've removed the barriers, I'd like to merge this?